### ISSUE TITLE
Concise and statically-typed `Property<T>` assignment

### ISSUE THREAD

**Issue by eskatos** [gradle/gradle#9268]
## Status Quo
**Extension**
```java
public interface Extension {
    Property<String> getDescription();
}
public interface OtherExtension {
    Property<String> getDescription();
}
```

**Groovy DSL**
```groovy
extension {
    description = "My $project.description"
    description = otherExtension.description
    description = provider {
        // compute stuff lazily
    }
}
```

**Kotlin DSL (build script or precompiled script)**
```kotlin
extension {
    description.set("My $project.description")
    description.set(otherExtension.description)
    description.set(provider {
        // compute stuff lazily
    })
}
```

### Option 1: Kotlin supports equals `=` operator overloading

**Implementation**
```kotlin
operator fun <T> Property<T>.EQUALS(value: T): Unit = set(value)
operator fun <T> Property<T>.EQUALS(provider: Provider<T>): Unit = set(provider)
```

**Kotlin DSL (build script or precompiled script)**
```kotlin
extension {
    description = "My $project.description"
    description = otherExtension.description
    description = provider {
        // compute stuff lazily
    }
}
```

**Groovy DSL**
Unchanged

**Pros/Cons**

* (+) Simple, just add two kotlin extensions
* (+) Groovy and Kotlin DSL look the same
* (-) Kotlin support not likely to happen

### Option 2: Rename `Property.set`/`get` to `setValue`/`getValue`

Just the for the plain value, do not rename `set(Provider<T>)`.

**Kotlin DSL (build script or precompiled script)**
```kotlin
extension {
    description.set("My $project.description") // DEPRECATED
    description.value = "My $project.description"
    description.set(otherExtension.description)
    description.set(provider {
        // compute stuff lazily
    })
}
```

**Groovy DSL**
```groovy
extension {
    description = "My $project.description" // DEPRECATED
    description.value = "My $project.description"
    description = otherExtension.description
}
```

**Pros/Cons**

* (+) Simple, just add two methods, deprecate two old ones
* (+) Explicit difference between using the plain value or connecting to other properties
* (-/+) Will not look the same in Groovy and Kotlin DSL unless we remove sugar from Groovy DSL
* (-) Looks nice for eager assignment only

**Alternative**

Also rename `set(Provider<T>)` to `setValue(Provider<T>)`.

```kotlin
extension {
    description.set("My $project.description") // DEPRECATED
    description.value = "My $project.description"
    description.setValue(otherExtension.description)
    description.setValue(provider {
        // compute stuff lazily
    })
}
```

This makes it worse at first until Kotlin supports overloaded property setters https://github.com/gradle/kotlin-dsl/issues/380

BUT it becomes semantically wrong, `value = provider` doesn't make much sense.

### Option 3: Kotlin invoke operator overloading

**Implementation**
```
operator fun <T> Property<T>.invoke(value: T?): Unit = set(value)
operator fun <T> Property<T>.invoke(provider: Provider<T>): Unit = set(provider)
```

**Kotlin DSL (build script or precompiled script)**
```kotlin
extension {
    description("foo")
    description(otherExtension.description)
    description(provider {
        // compute stuff lazily
    })
}
```

**Pros/Cons**

* (+) Simple, just add two kotlin extensions
* (-) Still not assignments

**Comment by Vampire** [comment]
It probably is out of the question as it will not strive to have the same syntax in the different languages, but I'd like to mention, that you could actually overload some compound-assignment operator in Kotlin like

```kotlin
operator fun <T> Property<T>.remAssign(value: T) = set(value)
operator fun <T> Property<T>.divAssign(value: T) = set(value)
operator fun <T> Property<T>.timesAssign(value: T) = set(value)
operator fun <T> Property<T>.invoke(value: T) = set(value)

tasks.githubPublish {
    prerelease.set(true)
    repositoryName %= githubRepositoryName
    tagName /= releaseTagName
    releaseName *= releaseTagName
    draft(true)
}
```

**Comment by SUPERCILEX** [comment]
Would option 1 be possible with a compiler plugin instead of operator overloads? So `=` would get rewritten to `.set` in the bytecode.

**Comment by sellmair** [comment]
I would like to also share something I have seen in the [kotlinx.coroutines](https://github.com/Kotlin/kotlinx.coroutines/blob/master/buildSrc/src/main/kotlin/MavenCentral.kt) repository: 
![image](https://user-images.githubusercontent.com/34319766/88274113-81742c80-ccdb-11ea-945c-e009c172a9ad.png)

**Comment by paplorinc** [comment]
Hey @eskatos, I see there wasn't really any activity on this issue for 2 years - any progress since?

**Comment by sdeleuze** [comment]
I recently hit this issue as well as we are starting to try using Dokka 1.4 in Spring Framework, and I think this one is critical for an healthy Gradle ecosystem. No strong opinion on the right solution, but a big :+1: from Spring team.

**Comment by eskatos** [comment]
@paplorinc nothing changed so far, the status quo in the issue description still applies

**Comment by sellmair** [comment]
What about providing a simple infix function as proposed above for now. This will already look and feel much better and can still be enhanced with a proper assignment operator later (once this becomes possible)?

**Comment by Vampire** [comment]
There is nothing wrong with you declaring this infix function for the projects where you want to use it or even provide it as library. But it would go the opposite direction. It would further separate Groovy and Kotlin syntaxes and with no real gain, just another different way to do the same thing while still not reaching consistency. So what good would that infix function does?

**Comment by sellmair** [comment]
I mean, as dokka contributor I could easily provide an infix fun like proposed with the plugin itself. But then maybe other plugins will do the same and we end up with many implementations. *This* will be confusing and we discussed this with the Gradle team and were explicitly advised not to do so. I think that "Kotlin DSL" users care about this `.set()` method much more than some divergence towards the Groovy syntax. Actually, this is one of the reasons I *want* to use Kotlin over Groovy.

**Comment by mpetuska** [comment]
I wonder why nobody mentioned kotlin delegate extension overloading? Seems like the most straightforward solution. The only issue with this might be tracking underlying Provider dependency chains.
```
operator fun <T> Provider<T>.getValue(thisRef: Any?, property: KProperty<*>): T {
  return get()
}

operator fun <T> Property<T>.setValue(thisRef: Any?, property: KProperty<*>, value: T) {
  return set(value)
}

fun usage() {
  val project = ProjectBuilder.builder().build()
  val provider by project.provider { 1 }
  println(provider + 1)
  var property by project.objects.property(Int::class.java)
  property = 69
  println(property)
  property = 420
  println(property)
}
```

**Comment by JLLeitschuh** [comment]
> The only issue with this might be tracking underlying Provider dependency chains.

This, I believe, is this biggest reason this API proposal doesn't work, unfortunately

**Comment by mpetuska** [comment]
@JLLeitschuh yep, but it would work with the second proposed option. This is less convenient from plugin author's perspective, but at least end-user DSL becomes smooth. Plus this pattern not that uncommon in kotlin, used a lot when mutable fields needs to be exposed as immutable to the outside.

**Comment by jamesward** [comment]
As a Gradle user the differences are painful.  How do I convert this block of Groovy to kts?  Do I need to call `.set` on this thing or use `=`?  Why sometimes is it one and sometimes the other?  Trial and error seem to be the frustrating strategy I use.  Yet just hiding the differences also hides the reasons why there are differences.  Presumably there is a reason why everything that can be set, is not a `Property<T>`.  When something is lazy, it also means that I do not know when it will be read, causing a whole different pain point.

For me the biggest underlying issue is a lack of examples which give me the right syntax.  Instead of changing the syntax again I think we should work more on updating StackOverflow answers, plugin docs, open source projects, and do whatever is needed to get the world from Groovy to kts.  I believe the pain identified by this issue mostly goes away when we get more of the world on kts.

**Comment by lptr** [comment]
> Presumably there is a reason why everything that can be set, is not a `Property<T>`.

Sadly, this is, by and large, not true. Well, there is a reason (we never finished the migration to the Provider API), but it's not a good one. We are actively exploring ways to convert every DSL-accessible property into a `Property` or `Provider` or some other lazy API without breaking the world as we speak. (The reason this issue come up recently is exactly because of this very effort.)

So hopefully in the near future every property accessible via the build DSL (Groovy or Kotlin) will be a lazy one. With that the meaning of `=` can also have a consistent meaning for these properties.

Alternatively, if we don't implement some way to use `=` to automatically convert to `.set()`, then all Kotlin DSL builds scripts will have to be converted to call `.set()` themselves, everywhere. This is both a lot of work, and will make the Groovy and Kotlin build files look very different. This might be a personal preference, but for me the syntax using `=` in Groovy is much preferable for its clarity.

What we would like to do is to make `=` work with properties of type `Property` and friends. This is significantly different than introducing a feature to Kotlin like general operator overloading. And while this still introduces some divergence from Kotlin the programming language, for Gradle's Kotlin DSL this might be a good compromise.

**Comment by mpetuska** [comment]
Or alternatively, drop `=` support in groovy and have both sides using `.set()` 🤔

**Comment by JavierSegoviaCordoba** [comment]
I would prefer a really big breaking change in Gradle 8-10 where the old approach is completely removed and the new one is added as default.

**Comment by bamboo** [comment]
> What we would like to do is to make = work with properties of type Property and friends.
I don't think this is settled yet.

**Comment by Goooler** [comment]
Might wait for https://github.com/Kotlin/KEEP/issues/309?

**Comment by asodja** [comment]
Overloading the `=` operator as described in the KEEP probably won't be a language feature, at least in a short term.

But we are working on the Kotlin compiler plugin, that will support overloading `=` for Kotlin Gradle DSL. It was not possible to write a plugin till recently since it required some changes to the Kotlin compiler. So if everything goes according to plan we will ship the experimental opt-in assignment for `Property<T>` in some 8.x version.

**Comment by asodja** [comment]
Current state for Kotlin assignment:
- Kotlin plugin integration was done in https://github.com/gradle/gradle/pull/22240 and is merged on master. 
- `=` for Property type for Kotlin can be enabled with `systemProp.org.gradle.unsafe.kotlin.assignment=true`. To make it work with IntelliJ this property has to be in the project root `gradle.properties` file.
- Dogfood PR for gradle/gradle is in https://github.com/gradle/gradle/pull/23547 

We found some issues with IntelliJ integrations that have to be addressed in IntelliJ:
1. Cannot use "Go to declaration" for properties on Java types when = is used (for Kotlin types it works ok), e.g.:
<img width="720" alt="212665996-ecca2462-0a91-43e8-82af-cedb37e02347" src="https://user-images.githubusercontent.com/3939893/213476709-f693c44d-4206-4433-97d8-cce15a6c178e.png">

2. IntelliJ incorrectly shows compile error for properties on extensions when using `<extension>.<property> = <value>` notation, e.g.:
<img width="592" alt="212665761-2bc0e64e-6c3a-4946-bf7b-09082cf4d2ae" src="https://user-images.githubusercontent.com/3939893/213476983-635af081-43a4-44cb-9682-a366fc912b32.png">

We might remove the opt-in flag if all the issues are resolved before 8.1, otherwise, we might remove it in 8.2.
